File: cpp/ahocorasick.cpp
#include <iostream>
#include <ostream>
#include <queue>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class TrieNode {
public:
    TrieNode *child[26] = {nullptr};
    TrieNode *output = nullptr;
    TrieNode *parent = nullptr;
    TrieNode *suffix = nullptr;
    string word;
    ~TrieNode() { for (auto& node : child) delete node; }
};


class AhoCorasick {
public:
    TrieNode *root = new TrieNode();

    ~AhoCorasick() { delete root; }

    void build(vector<string> patterns) {
        for (auto& pattern : patterns) {
            TrieNode *node = root;
            for (auto& ch : pattern) {
                int c = ch - 'a';
                if (!node->child[c]) {
                    node->child[c] = new TrieNode();
                    node->child[c]->parent = node;
                }
                node = node->child[c];
            }
            node->word = pattern;
        }
        queue<TrieNode*> q; q.push(root);
        while (q.size())
            for (int sz = q.size(); sz; --sz) {
                TrieNode *node = q.front(); q.pop();
                for (int i = 0; i < 26; ++i) {
                    TrieNode *child = node->child[i], *suffix = node->suffix;
                    if (child) {
                        while (suffix && suffix->child[i] == nullptr)
                            suffix = suffix->suffix;
                        if (suffix) {
                            child->suffix = suffix->child[i];
                            if (child->suffix->word != "") child->output = child->suffix;
                            else child->output =child->suffix->output;
                        } else {
                            child->output = nullptr;
                            child->suffix = root;
                        }
                        q.push(child);
                    }
                }
            }
    }

    unordered_map<string, vector<int>> match(string text) {
        unordered_map<string, vector<int>> ans;
        TrieNode *node = root;
        for (int i = 0; i < text.size(); ++i) {
            char ch = text[i];
            int c = ch - 'a';
            while (!node->child[c] && node->suffix) node = node->suffix;
            if (node->child[c]) node = node->child[c];
            for (TrieNode *output = node; output; output = output->output)
                if (!output->word.empty()) {
                    string pattern = output->word;
                    ans[pattern].push_back(i-pattern.size()+1);
                }
        }
        return ans;
    }
};


int main() {
    AhoCorasick *trie = new AhoCorasick();
    vector<string> words = {"abc", "aaaaa", "bcdef"};
    trie->build(words);
    unordered_map<string, vector<int>> ans = trie->match("aaaaaaabcdef");
    for (auto& [k, v] : ans) {
        cout << k << " : ";
        for (auto& x : v)
            cout << x << " ";
        cout << endl;
    }
    return 0;
}

File: java/ahocorasick.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Random;
import java.util.Stack;

class TrieNode {
    public TrieNode[] child = new TrieNode[26];
    public TrieNode output = null;
    public TrieNode parent = null;
    public TrieNode suffix = null;
    public String word;
}


public class AhoCorasick {
    public TrieNode root = new TrieNode();

    public void build(String[] patterns) {
        for (var pattern : patterns) {
            TrieNode node = root;
            for (var ch : pattern.toCharArray()) {
                int c = (int) (ch - 'a');
                if (node.child[c] == null) {
                    node.child[c] = new TrieNode();
                    node.child[c].parent = node;
                }
                node = node.child[c];
            }
            node.word = pattern;
        }
        Queue<TrieNode> q = new LinkedList<>(); q.add(root);
        while (!q.isEmpty())
            for (int sz = q.size(); sz > 0; --sz) {
                TrieNode node = q.poll();
                for (int i = 0; i < 26; ++i) {
                    TrieNode child = node.child[i], suffix = node.suffix;
                    if (child != null) {
                        while (suffix != null && suffix.child[i] == null)
                            suffix = suffix.suffix;
                        if (suffix != null) {
                            child.suffix = suffix.child[i];
                            if (child.suffix.word != null) child.output =child.suffix;
                            else child.output = child.suffix.output;
                        } else {
                            child.output = null;
                            child.suffix = root;
                        }
                        q.add(child);
                    }
                }
            }
    }

    public Map<String, List<Integer>> match(String text) {
        Map<String, List<Integer>> ans = new HashMap<>();
        TrieNode node = root;
        for (int i = 0; i < text.length(); ++i) {
            char ch = text.charAt(i);
            int c = (int) (ch - 'a');
            while (node.child[c] == null && node.suffix != null)
                node = node.suffix;
            if (node.child[c] != null) node = node.child[c];
            for (TrieNode output = node; output != null; output = output.output)
                if (output.word != null) {
                    String pattern = output.word;
                    if (!ans.containsKey(pattern))
                        ans.put(pattern, new ArrayList<>());
                    ans.get(pattern).add(i-pattern.length()+1);
                }
        }
        return ans;
    }

    public static void main(String[] args) {
        AhoCorasick trie = new AhoCorasick();
        List<String> words = new ArrayList<>(Arrays.asList("abc", "aaaaa", "bcdef"));
        trie.build(words);
        Map<String, List<Integer>> ans = trie.match("aaaaaaabcdef");
        for (var k : ans.keySet()) {
            System.out.print(k + " : ");
            for (var v : ans.get(k))
                System.out.print(v + " ");
            System.out.println();
        }
    }
}
